### PR TITLE
Implement MySQL, PostgreSQL, Redshift Connectors' pagination

### PR DESCRIPTION
**Pull request by ejeffrli**
Add postgresql and redshift pagination implementations

*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by akuzin1 — APPROVED** [review]

File: athena-mysql/src/main/java/com/amazonaws/athena/connectors/mysql/MySqlMetadataHandler.java [hunk 1]
```diff
@@ -215,11 +218,47 @@ public GetSplitsResponse doGetSplits(
...
+    }
+
+    @Override
+    protected ListTablesResponse listPaginatedTables(final Connection connection, final ListTablesRequest listTablesRequest) throws SQLException
+    {
+        String token = listTablesRequest.getNextToken();
+        int pageSize = listTablesRequest.getPageSize();
+
+        // no pagination case
+        if (pageSize == UNLIMITED_PAGE_SIZE_VALUE && token == null) {
```
Comment (chngpe) [inline athena-mysql/src/main/java/com/amazonaws/athena/connectors/mysql/MySqlMetadataHandler.java:240]: This segment of code id duplicated and this logic is common regardless of JDBC source type, can we move it 1 layer up as common logic in doListTables?

File: athena-mysql/src/main/java/com/amazonaws/athena/connectors/mysql/MySqlMetadataHandler.java [hunk 2]
```diff
@@ -215,11 +218,47 @@ public GetSplitsResponse doGetSplits(
...
+        }
+
+        // pagination case
+        int t = 0;
+        if (token != null) {
+            t = Integer.valueOf(token);
+        }
+        List<TableName> paginatedTables = getPaginatedTables(connection, listTablesRequest.getSchemaName(), t, pageSize);
+        LOGGER.info("{} tables returned from start token, {}, and end token, {}", paginatedTables.size(), Integer.toString(t), Integer.toString(t + pageSize));
+        ListTablesResponse response = new ListTablesResponse(listTablesRequest.getCatalogName(), paginatedTables, Integer.toString(t + pageSize));
```
Comment (chngpe) [inline athena-mysql/src/main/java/com/amazonaws/athena/connectors/mysql/MySqlMetadataHandler.java:251]: Can we add unit test or integration test to make sure we the token/pagination logic won't return duplicate? Ideally it should be like
Page size: 20
1st response: [0,20)
2n response: [20,40)

**Review by chngpe — CHANGES_REQUESTED** [review]

File: athena-postgresql/src/main/java/com/amazonaws/athena/connectors/postgresql/PostGreSqlMetadataHandler.java [hunk 3]
```diff
@@ -379,8 +405,21 @@ protected ArrowType getArrayArrowTypeFromTypeName(String typeName, int precision
         }
     }
 
+    protected List<TableName> getPaginatedResults(Connection connection, String databaseName, int token, int limit) throws SQLException
+    {
+        String tablesAndViews = "Tables and Views";
+        String sql = "SELECT a.\"TABLE_NAME\", a.\"TABLE_SCHEM\" FROM ((SELECT table_name as \"TABLE_NAME\", table_schema as \"TABLE_SCHEM\" FROM information_schema.tables WHERE table_schema = ?) UNION (SELECT matviewname as \"TABLE_NAME\", schemaname as \"TABLE_SCHEM\" from pg_catalog.pg_matviews mv where has_table_privilege(format('%I.%I', mv.schemaname, mv.matviewname), 'select') and schemaname = ?)) AS a ORDER BY a.\"TABLE_NAME\" LIMIT ? OFFSET ?";
```
Comment (chngpe) [inline athena-postgresql/src/main/java/com/amazonaws/athena/connectors/postgresql/PostGreSqlMetadataHandler.java:411]: move this to static

File: athena-redshift/src/main/java/com/amazonaws/athena/connectors/redshift/RedshiftMetadataHandler.java [hunk 4]
```diff
@@ -124,4 +125,17 @@ public GetDataSourceCapabilitiesResponse doGetDataSourceCapabilities(BlockAlloca
 
         return new GetDataSourceCapabilitiesResponse(request.getCatalogName(), capabilities.build());
     }
+
+    @Override
+    protected List<TableName> getPaginatedResults(Connection connection, String databaseName, int token, int limit) throws SQLException
+    {
+        String tablesAndViews = "Tables and Views";
+        String sql = "SELECT table_name as \"TABLE_NAME\", table_schema as \"TABLE_SCHEM\" FROM information_schema.tables WHERE table_schema = ? ORDER BY \"TABLE_NAME\" LIMIT ? OFFSET ?";
```
Comment (chngpe) [inline athena-redshift/src/main/java/com/amazonaws/athena/connectors/redshift/RedshiftMetadataHandler.java:133]: static var please

File: athena-postgresql/src/main/java/com/amazonaws/athena/connectors/postgresql/PostGreSqlMetadataHandler.java (hunk 3 above)
Comment (ejeffrli) [reply athena-postgresql/src/main/java/com/amazonaws/athena/connectors/postgresql/PostGreSqlMetadataHandler.java:411]: the sql variable or tablesandViews variable?

File: athena-redshift/src/main/java/com/amazonaws/athena/connectors/redshift/RedshiftMetadataHandler.java [hunk 5]
```diff
@@ -124,4 +128,15 @@ public GetDataSourceCapabilitiesResponse doGetDataSourceCapabilities(BlockAlloca
 
         return new GetDataSourceCapabilitiesResponse(request.getCatalogName(), capabilities.build());
     }
+
+    @Override
+    protected List<TableName> getPaginatedResults(Connection connection, String databaseName, int token, int limit) throws SQLException
```
[the remark this answers is not in the document]
Comment (ejeffrli) [reply athena-redshift/src/main/java/com/amazonaws/athena/connectors/redshift/RedshiftMetadataHandler.java:133]: It works with materialized views. The previous implementation calls getMaterializedViews(), but it lets it error out because pg_catalog.pg_matviews is not supported by redshift. Pagination still keeps the essential part of the query.
